### ISSUE TITLE
how to run the optmization for a fixed number of epochs ?

### ISSUE THREAD

**Issue by rodrigob** [lmjohns3/downhill#6]
from the documentation there seem to be no way to fix the number of optimization iterations (the documentation only describes stopping conditions based on the results, not on the number of iterations).

**Comment by lmjohns3** [comment]
Yes, downhill uses early stopping to determine when to halt training.

If you really want to run for a fixed number of iterations ``N``, use these parameter settings:

    patience=0
    min_improvement=0.9999
    validate_every=N

This will run one validation after ``N`` updates, which will fail (unless somehow the validation error is actually 99.99% better than the initial state), and then training will halt because the patience will be exhausted.

**Comment by rodrigob** [comment]
so min_improvement should be `float("inf")` to be really sure it will stop...

**Comment by lmjohns3** [comment]
Yes, actually, any value >= 1 will work, too. The relevant line (https://github.com/lmjohns3/downhill/blob/master/downhill/base.py#L265) is:

        if self._best_loss - loss > self._best_loss * self.min_improvement:

So the comparison is between the previous best and the current validation loss; this quantity is limited to the range ``[-inf, self._best_loss]`` as long as the loss is nonnegative, so anything 1 or greater for ``min_improvement`` should do the trick.

**Comment by rodrigob** [comment]
But I notice that using this strategy the params are not set `self._best_params` at the end.
Using this from `dowhill.minimize` then gives no access to the optimization result !

**Comment by lmjohns3** [comment]
Hm, yes, I see how this is a problem.

It might be easiest if I just add a ``max_updates`` parameter. To use it, make sure ``patience`` is set to a large enough value to guarantee that ``max_updates`` iterations will take place during optimization. The rest of the validation infrastructure will keep working as it currently does.
